From: Oleg Nesterov <oleg@redhat.com>
Subject: document while_each_thread(), change first_tid() to use for_each_thread()
Date: Wed, 23 Aug 2023 19:08:06 +0200

Add the comment to explain that while_each_thread(g,t) is not rcu-safe
unless g is stable (e.g.  current).  Even if g is a group leader and thus
can't exit before t, t or another sub-thread can exec and remove g from
the thread_group list.

The only lockless user of while_each_thread() is first_tid() and it is
fine in that it can't loop forever, yet for_each_thread() looks better and
I am going to change while_each_thread/next_thread.

Link: https://lkml.kernel.org/r/20230823170806.GA11724@redhat.com
Signed-off-by: Oleg Nesterov <oleg@redhat.com>
Cc: Eric W. Biederman <ebiederm@xmission.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 fs/proc/base.c               |    5 ++---
 include/linux/sched/signal.h |    4 ++++
 2 files changed, 6 insertions(+), 3 deletions(-)

--- a/fs/proc/base.c~document-while_each_thread-change-first_tid-to-use-for_each_thread
+++ a/fs/proc/base.c
@@ -3813,11 +3813,10 @@ static struct task_struct *first_tid(str
 	/* If we haven't found our starting place yet start
 	 * with the leader and walk nr threads forward.
 	 */
-	pos = task = task->group_leader;
-	do {
+	for_each_thread(task, pos) {
 		if (!nr--)
 			goto found;
-	} while_each_thread(task, pos);
+	};
 fail:
 	pos = NULL;
 	goto out;
--- a/include/linux/sched/signal.h~document-while_each_thread-change-first_tid-to-use-for_each_thread
+++ a/include/linux/sched/signal.h
@@ -648,6 +648,10 @@ extern void flush_itimer_signals(void);
 
 extern bool current_is_single_threaded(void);
 
+/*
+ * Without tasklist/siglock it is only rcu-safe if g can't exit/exec,
+ * otherwise next_thread(t) will never reach g after list_del_rcu(g).
+ */
 #define while_each_thread(g, t) \
 	while ((t = next_thread(t)) != g)
 
_
